phpBB/phpbb/mention/helper/mention_helper.php

```diff
@@ -0,0 +1,123 @@
+<?php
+/**
+ *
+ * phpBB mentions. A feature for the phpBB Forum Software package.
+ *
+ * @copyright (c) 2016, paul999, https://www.phpbbextensions.io
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ */
+
+namespace phpbb\mention\helper;
+use phpbb\db\driver\driver_interface;
+use phpbb\request\request_interface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use phpbb\controller\helper;
+
+
+class mention_helper {
+	/**
+	* @var driver_interface
+	*/
+	private $db;
+
+	/**
+	* @var Notification Details
+	*/
+	private $data;
+
+	public function __construct(driver_interface $db) {
+		$this->db = $db;
+	}
+
+	public function get_mentioned_users($post_text, $data, $notif_manager_obj) {
+		$this->data = $data;
+		$regular_expression_match = '#\[mention\](.*?)\[/mention\]#';
+		$matches = false;
+		$user_list = $this->get_userlist();
+		preg_match_all($regular_expression_match, $post_text, $matches, PREG_OFFSET_CAPTURE);
+		$start_tag_length = strlen("[mention]");
+		$end_tag_length = strlen("[\mention]");
+		$users_notified = array();
+		for ($i = 0, $len = count($matches[1]); $i < $len; $i++)
+		{
+			$username_clean = utf8_clean_string($matches[1][$i][0]);
+			if (array_key_exists($username_clean, $user_list))
+			{
+				if(!in_array($username_clean, $users_notified, true)){
+
+					$startpos = $matches[1][$i][1] - $start_tag_length;
+					$length = strlen($matches[1][$i][0]);
+					$endpos = $matches[1][$i][1] + $length + $end_tag_length - 1;
+					$user_data = $user_list[$username_clean];
+					$add_url_tag = "[url= memberlist.php?mode=viewprofile&u=".$user_data."]".$username_clean."[/url]";
+					if ($i==0) {
+						$new_post_text = substr($post_text, 0, $startpos);
+						$new_post_text = $new_post_text . $add_url_tag;
+						$prev_end_pos = $endpos + 1;
+					}else {
+						$strip_mention = substr($post_text, $prev_end_pos, $startpos - $prev_end_pos);
+						$new_post_text = $new_post_text . $strip_mention;
+						$new_post_text = $new_post_text . $add_url_tag;
+						$prev_end_pos = $endpos + 1;
+					}
+					$this->send_notifications($user_list[$username_clean], $notif_manager_obj);
```

Comment:
Be carefull, the send_notifications causes at least 1 query for every time it is ran. Depending how many users are notified in a post this might cause a really high number of queries.